lxd/daemon.go

```diff
@@ -1322,3 +1322,50 @@ func initializeDbObject(d *Daemon) (*db.Dump, error) {
 
 	return dump, nil
 }
+
+func (d *Daemon) hasNodeListChanged(heartbeatData *cluster.APIHeartbeat) bool {
+	// No previous heartbeat data.
+	if d.lastNodeList == nil {
+		return true
+	}
+
+	// Member count has changed.
+	if len(d.lastNodeList.Members) != len(heartbeatData.Members) {
+		return true
+	}
+
+	// Check for node address changes.
+	for lastMemberID, lastMember := range d.lastNodeList.Members {
+		if heartbeatData.Members[lastMemberID].Address != lastMember.Address {
+			return true
+		}
+	}
+
+	return false
+}
+
+// NodeRefreshTask is run each time a fresh node is generated. This can be used to trigger actions
+// when the node list changes.
+func (d *Daemon) NodeRefreshTask(heartbeatData *cluster.APIHeartbeat) {
+	for i, node := range heartbeatData.Members {
+		// Exclude nodes that haven't been heard from for a while.
+		// This is to avoid forkdns delaying results by querying an offline node.
+		if time.Since(node.LastHeartbeat) > (time.Second * 30) {
```

Comment:
@stgraber @freeekanayaka does 30s seem like a good time to consider a node offline/unavailable?

Reply:
@stgraber @freeekanayaka I went for 3*HeartbeatInterval in the end.